stock_picking_propagate_scheduled_date/models/stock_move.py

```diff
@@ -2,32 +2,80 @@
 
 
 class StockMove(models.Model):
     _inherit = "stock.move"
 
-    def _propagate_date(self, new_date):
-        """Propagate the date of a move to its destination."""
-        already_propagated_ids = self.env.context.get(
-            "date_propagation_ids", set()
-        ) | set(self.ids)
-        self = self.with_context(date_propagation_ids=already_propagated_ids)
-        for move in self:
-            if move.date:
-                delta = move.date - fields.Datetime.to_datetime(new_date)
-            else:
-                delta = 0
-            for move_dest in move.move_dest_ids:
-                if move_dest.state in ("done", "cancel"):
-                    continue
-                if move_dest.id in already_propagated_ids:
-                    continue
-                move_dest.date -= delta
-
     def write(self, vals):
-        # propagate date changes in the stock move chain
-        if "date" in vals:
-            self._propagate_date(vals.get("date"))
-        return super().write(vals)
+        """Override the write method to propagate the date to the destination moves."""
+        need_propagate = "date" in vals and not self.env.context.get(
+            "_propagate_date_no_recursion"
+        )
+        if need_propagate:
+            # Get the previous date of the moves
+            previous_dates = {
+                move.id: move.date for move in self.with_context(prefetch_fields=False)
+            }
+        res = super().write(vals)
+        if need_propagate:
+            self._propagate_date_no_recursion(previous_dates, vals.get("date"))
+        return res
+
+    def _get_moves_in_down_chain(self):
+        """Get all the destination moves in the down chain of a moves recordset recursively."""
+        query = """
+            WITH RECURSIVE chained_moves AS (
+                SELECT
+                    r.move_orig_id, r.move_orig_id as from_id, r.move_dest_id as dest_id
+                FROM
+                    stock_move_move_rel r
+                WHERE r.move_orig_id IN %s
+
+                UNION ALL
+
+                SELECT
+                    chained_moves.move_orig_id,
+                    r.move_orig_id as from_id,
+                    r.move_dest_id as dest_id
+                FROM
+                    stock_move_move_rel r
+                JOIN chained_moves ON r.move_orig_id = chained_moves.dest_id
+            )
+
+            SELECT
+                move_orig_id,
+                array_agg(dest_id)
+            FROM chained_moves
+            GROUP BY move_orig_id;
+        """
+
+        self.env.cr.execute(query, (tuple(self.ids),))
+        # key: original move, values: list of destination moves
+        chained = dict(self.env.cr.fetchall())
+        return chained
+
+    def _propagate_date_no_recursion(self, previous_dates, new_date):
+        """Propagate the date of a move to its destination avoiding recursion."""
+        if not self:
+            return True
```

Comment:
No need to return something:
```suggestion
            return
```